feat(grouping): Added support for grouping components

mitsuhiko opened the pull request:
This pull request adds support for grouping components. The goal of it is that meta information can be provided about why groups are being created. This PR does not yet fully replace the normal hash generation but it expresses `calculate_hashes` by using the components internally.

The longer term goal is to replace all of `calculate_hashes` on the interfaces and to move the current implementation of the grouping code into a default strategy so we can provide multiple implementations of it.

Rough overview of how this works now:

### Terms

We now have a few new terms to describe grouping. The most important ones are these:

* `GroupingComponent`: a grouping component is a single node in a tree of contributing nodes to a grouping hash. When the hash is produced for a grouping component it is flattened into a list and ultimately joined into an md5 hash. Each component has a bool that indicates if a component contributes to the hash (`contributes`) which is checked recursively. It has an `id` which gives a basic id that indicates which type of component it is. It also holds a list of `values` which are either primitives that are hashed or recursive references to other grouping components.
* a **grouping variant**: when an interface tries to produce a grouping component it will do so multiple times once for each variant. Currently there are two variants: `system` and `app`. Per policy each interface needs to implement `system` and `app` can be supplied optionally if only in-app values should be considered.

### Grouping Variant Reporting

Because of checksums and fingerprints grouping variants are reported out of the event in a slightly different form than the normal grouping code would suggest. If no fingerprint is set and no checksum (or the fingerprint is just `{{ default }}`) then the variants reported are most likely `app` and `system` as one suggests. However when checksum are involved the reported variants are instead `checksum` and `hashed-checksum`. Likewise if a fingerprint does not involve `{{ default }}` at all a single variant called `custom-fingerprint` is reported. Otherwise one can again expect `app` and `system` but additional values are salted in.

The idea is to expose the grouping variants to the UI for which the UI layer needs to understand how to interpret the different types of grouping variants.

### Example Dump

This is the example dump from one of the test cases:

<details>
  <summary>Click to show example dump</summary>

```json
{
  "app": {
    "component": {
      "contributes": true,
      "values": [
        {
          "contributes": true,
          "values": [
            {
              "contributes": true,
              "values": [
                {
                  "contributes": true,
                  "values": [
                    {
                      "contributes": true,
                      "values": [
                        {
                          "contributes": false,
                          "values": [],
                          "id": "module",
                          "hint": null
                        },
                        {
                          "contributes": true,
                          "values": [
                            "foo/baz.py"
                          ],
                          "id": "filename",
                          "hint": null
                        },
                        {
                          "contributes": false,
                          "values": [],
                          "id": "context-line",
                          "hint": null
                        },
                        {
                          "contributes": false,
                          "values": [],
                          "id": "symbol",
                          "hint": null
                        },
                        {
                          "contributes": false,
                          "values": [],
                          "id": "function",
                          "hint": null
                        },
                        {
                          "contributes": true,
                          "values": [
                            1
                          ],
                          "id": "lineno",
                          "hint": null
                        }
                      ],
                      "id": "frame",
                      "hint": null
                    }
                  ],
                  "id": "stacktrace",
                  "hint": null
                },
                {
                  "contributes": true,
                  "values": [
                    "ValueError"
                  ],
                  "id": "type",
                  "hint": null
                },
                {
                  "contributes": false,
                  "values": [],
                  "id": "value",
                  "hint": "stacktrace and type take precedence"
                }
              ],
              "id": "exception",
              "hint": null
            },
            {
              "contributes": false,
              "values": [
                {
                  "contributes": false,
                  "values": [
                    {
                      "contributes": false,
                      "values": [
                        {
                          "contributes": false,
                          "values": [],
                          "id": "module",
                          "hint": null
                        },
                        {
                          "contributes": true,
                          "values": [
                            "foo/baz.py"
                          ],
                          "id": "filename",
                          "hint": null
                        },
                        {
                          "contributes": false,
                          "values": [],
                          "id": "context-line",
                          "hint": null
                        },
                        {
                          "contributes": false,
                          "values": [],
                          "id": "symbol",
                          "hint": null
                        },
                        {
                          "contributes": false,
                          "values": [],
                          "id": "function",
                          "hint": null
                        },
                        {
                          "contributes": true,
                          "values": [
                            1
                          ],
                          "id": "lineno",
                          "hint": null
                        }
                      ],
                      "id": "frame",
                      "hint": "non app frame"
                    }
                  ],
                  "id": "stacktrace",
                  "hint": "less than 10% of frames are in-app"
                },
                {
                  "contributes": false,
                  "values": [],
                  "id": "type",
                  "hint": "stacktrace takes precedence"
                },
                {
                  "contributes": false,
                  "values": [],
                  "id": "value",
                  "hint": "stacktrace takes precedence"
                }
              ],
              "id": "exception",
              "hint": null
            }
          ],
          "id": "exception",
          "hint": null
        }
      ],
      "id": "app",
      "hint": null
    },
    "hash": "baf0bfa8bb312627cd5ef4171aec9f4e",
    "type": "component"
  },
  "system": {
    "component": {
      "contributes": true,
      "values": [
        {
          "contributes": true,
          "values": [
            {
              "contributes": true,
              "values": [
                {
                  "contributes": true,
                  "values": [
                    {
                      "contributes": true,
                      "values": [
                        {
                          "contributes": false,
                          "values": [],
                          "id": "module",
                          "hint": null
                        },
                        {
                          "contributes": true,
                          "values": [
                            "foo/baz.py"
                          ],
                          "id": "filename",
                          "hint": null
                        },
                        {
                          "contributes": false,
                          "values": [],
                          "id": "context-line",
                          "hint": null
                        },
                        {
                          "contributes": false,
                          "values": [],
                          "id": "symbol",
                          "hint": null
                        },
                        {
                          "contributes": false,
                          "values": [],
                          "id": "function",
                          "hint": null
                        },
                        {
                          "contributes": true,
                          "values": [
                            1
                          ],
                          "id": "lineno",
                          "hint": null
                        }
                      ],
                      "id": "frame",
                      "hint": null
                    }
                  ],
                  "id": "stacktrace",
                  "hint": null
                },
                {
                  "contributes": true,
                  "values": [
                    "ValueError"
                  ],
                  "id": "type",
                  "hint": null
                },
                {
                  "contributes": false,
                  "values": [],
                  "id": "value",
                  "hint": "stacktrace and type take precedence"
                }
              ],
              "id": "exception",
              "hint": null
            },
            {
              "contributes": true,
              "values": [
                {
                  "contributes": true,
                  "values": [
                    {
                      "contributes": true,
                      "values": [
                        {
                          "contributes": false,
                          "values": [],
                          "id": "module",
                          "hint": null
                        },
                        {
                          "contributes": true,
                          "values": [
                            "foo/baz.py"
                          ],
                          "id": "filename",
                          "hint": null
                        },
                        {
                          "contributes": false,
                          "values": [],
                          "id": "context-line",
                          "hint": null
                        },
                        {
                          "contributes": false,
                          "values": [],
                          "id": "symbol",
                          "hint": null
                        },
                        {
                          "contributes": false,
                          "values": [],
                          "id": "function",
                          "hint": null
                        },
                        {
                          "contributes": true,
                          "values": [
                            1
                          ],
                          "id": "lineno",
                          "hint": null
                        }
                      ],
                      "id": "frame",
                      "hint": null
                    }
                  ],
                  "id": "stacktrace",
                  "hint": null
                },
                {
                  "contributes": true,
                  "values": [
                    "ValueError"
                  ],
                  "id": "type",
                  "hint": null
                },
                {
                  "contributes": false,
                  "values": [],
                  "id": "value",
                  "hint": "stacktrace and type take precedence"
                }
              ],
              "id": "exception",
              "hint": null
            }
          ],
          "id": "exception",
          "hint": null
        },
        {
          "contributes": false,
          "values": [
            "message"
          ],
          "id": "message",
          "hint": "exception strategy takes precedence"
        }
      ],
      "id": "system",
      "hint": null
    },
    "hash": "9aef9026a457d312531c07644b271314",
    "type": "component"
  }
}
```

</details>

mitsuhiko commented:
Example dump of what this looks like for one of the testcases:

```json
{
  "contributes": true,
  "values": [
    {
      "contributes": true,
      "values": [
        {
          "contributes": true,
          "values": [
            {
              "contributes": true,
              "values": [
                {
                  "contributes": false,
                  "values": [],
                  "id": "module",
                  "hint": null
                },
                {
                  "contributes": true,
                  "values": [
                    "foo/baz.py"
                  ],
                  "id": "filename",
                  "hint": null
                },
                {
                  "contributes": false,
                  "values": [],
                  "id": "context-line",
                  "hint": null
                },
                {
                  "contributes": false,
                  "values": [],
                  "id": "symbol",
                  "hint": null
                },
                {
                  "contributes": false,
                  "values": [],
                  "id": "function",
                  "hint": null
                },
                {
                  "contributes": true,
                  "values": [
                    1
                  ],
                  "id": "lineno",
                  "hint": null
                }
              ],
              "id": "frame",
              "hint": null
            }
          ],
          "id": "stacktrace",
          "hint": null
        },
        {
          "contributes": true,
          "values": [
            "ValueError"
          ],
          "id": "type",
          "hint": null
        },
        {
          "contributes": false,
          "values": [],
          "id": "value",
          "hint": "value takes precedence"
        }
      ],
      "id": "exception",
      "hint": null
    },
    {
      "contributes": true,
      "values": [
        {
          "contributes": true,
          "values": [
            {
              "contributes": true,
              "values": [
                {
                  "contributes": false,
                  "values": [],
                  "id": "module",
                  "hint": null
                },
                {
                  "contributes": true,
                  "values": [
                    "foo/baz.py"
                  ],
                  "id": "filename",
                  "hint": null
                },
                {
                  "contributes": false,
                  "values": [],
                  "id": "context-line",
                  "hint": null
                },
                {
                  "contributes": false,
                  "values": [],
                  "id": "symbol",
                  "hint": null
                },
                {
                  "contributes": false,
                  "values": [],
                  "id": "function",
                  "hint": null
                },
                {
                  "contributes": true,
                  "values": [
                    1
                  ],
                  "id": "lineno",
                  "hint": null
                }
              ],
              "id": "frame",
              "hint": null
            }
          ],
          "id": "stacktrace",
          "hint": null
        },
        {
          "contributes": true,
          "values": [
            "ValueError"
          ],
          "id": "type",
          "hint": null
        },
        {
          "contributes": false,
          "values": [],
          "id": "value",
          "hint": "value takes precedence"
        }
      ],
      "id": "exception",
      "hint": null
    }
  ],
  "id": "exception",
  "hint": null
}
```

mitsuhiko commented:
I still need to figure out how to actually feed the variants and checksum into it so that I can use the components till the very end and expose that info on-demand for the UI.

mitsuhiko commented:
I think this is as far as I want to go with the current pull request. To actually get rid of `calculate_hashes` there is higher risk involved and I want to do this separately.

mitsuhiko commented:
This has at least one regression in master that is covered by tests but that I can reproduce locally. The behavior of exception grouping has slightly changed producing two hashes with this PR for one case where master has one.

![image](https://user-images.githubusercontent.com/7396/53703314-decebb80-3e10-11e9-8b35-5c3d0774bc4e.png)

mitsuhiko commented:
TODO: add a test for the case where the app variant is made for the exception but none of the frames are in-app. In that case the old grouping behavior switches to all frames are in-app. Without that behavior the grouping deteriorates for unknown reasons. This is currently not covered by a test.

tkaemming commented on src/sentry/interfaces/exception.py:
```diff
@@ -1036,30 +1066,39 @@ def to_json(self):
-                                                 variant=variant))
-
-        return output
+    def get_grouping_component(self, platform=None, variant=None):
```
note: Need to come back to this.

tkaemming commented on src/sentry/interfaces/stacktrace.py:
```diff
@@ -837,48 +931,61 @@ def to_json(self):
+            in_app_count = sum(1 if f.in_app else 0 for f in frames)
+            if in_app_count == 0:
+                in_app_count = total_frames
+                all_frames_are_in_app = True
```
The variable name `all_frames_are_in_app` is confusing and misleading when reading expressions like `not frame.in_app and not all_frames_are_in_app` below. Maybe `treat_ all_frames_as_in_app` or `condsider_all_frames_in_app` would be better?

mitsuhiko replied:
Indeed.

tkaemming commented on src/sentry/interfaces/stacktrace.py:
```diff
@@ -837,48 +931,61 @@ def to_json(self):
+                    contributes=False,
+                    hint='ignored due to recursion',
+                )
+            elif variant == 'app' and not frame.in_app and all_frames_are_in_app:
```
Kinda tricky to read this because it's not obvious the recursion check needs to be performed first.

tkaemming commented on src/sentry/interfaces/stacktrace.py:
```diff
@@ -837,48 +931,61 @@ def to_json(self):
-        if variant not in ('app', 'system'):
-            return []
-
+    def get_grouping_component(self, platform=None, variant=None):
```
Makes sense.

tkaemming commented on src/sentry/interfaces/template.py:
```diff
@@ -58,10 +59,22 @@ def to_python(cls, data):
-        if variant not in ('app', 'system'):
-            return []
-        return [self.filename, self.context_line]
+    def get_grouping_component(self, platform=None, variant=None):
```
Makes sense.

tkaemming commented on src/sentry/interfaces/stacktrace.py:
```diff
@@ -452,74 +473,146 @@ def to_json(self):
-        Our ideal scenario is the module name in addition to the line of
-        context. However, in several scenarios we opt for other approaches due
-        to platform constraints.
+    def get_grouping_component(self, platform=None, variant=None):
```
note: Need to come back to this.

tkaemming commented on src/sentry/interfaces/threads.py:
```diff
@@ -86,12 +88,25 @@ def get_meta_context(self, meta, is_public=False):
-    def get_hash(self, platform=None, variant='system'):
-        if variant not in ('app', 'system'):
-            return []
+    def get_grouping_component(self, platform=None, variant=None):
```
This one is easy. 😄

mitsuhiko commented on src/sentry/event_hashing.py:
```diff
@@ -11,13 +11,292 @@
+        self.values = values
+
+    def get_subcomponent(self, id):
+        """Looks up a subcomponent by the id and returns the first or `None`."""
```
Used exclusively to find the stacktrace on exceptions to replicate our old chained exception behavior.

mitsuhiko commented on src/sentry/event_hashing.py:
```diff
@@ -11,13 +11,292 @@
+            else:
+                # this basically assumes that a value is only a primitive
+                # and never an object or list.  This should be okay
+                # because we verify this.
```
good point, lost the assert in the ctor at one point. Will add again.

mitsuhiko commented on src/sentry/interfaces/exception.py:
```diff
@@ -1036,30 +1066,39 @@ def to_json(self):
-        # while others may not and we ALWAYS want stacktraces over values
-        output = []
-        for value in self._values():
-            if not value or not value.stacktrace:
```
values is implemented to skip nones now. So I guess that code path got stale.

mitsuhiko commented on src/sentry/event_hashing.py:
```diff
@@ -11,13 +11,292 @@
+    def __init__(self, id, hint=None, contributes=None, values=None):
+        self.id = id
+        if hint is None:
+            hint = DEFAULT_HINTS.get(id)
```
I wanted to refactor this further by instead of passing `id` i pass instances of preconfigured GroupTypes or something. Then it gets clearer. On the UI branch I also now derive names from this so one needs to know that the IDs are not just entirely arbitrary. I think that will clarify this a little bit.

So it might turn into something like this:

```python
salt = GroupType(id='salt', hint='a static salt')
GroupingComponent(ty=salt, values=['...'])
```

mitsuhiko commented on src/sentry/event_hashing.py:
```diff
@@ -11,13 +11,292 @@
+        return self.hash
+
+
+class ComponentVariant(BaseVariant):
```
This is actually clearer to see in the other PR that builds on this: #12270. The output from `as_dict` is used by the UI to render it and it uses the type to dispatch the rendering.

mitsuhiko commented on src/sentry/event_hashing.py:
```diff
@@ -11,13 +11,292 @@
+
+
+class BaseVariant(object):
+    type = None
```
This is used for the grouping UI code in #12270

mitsuhiko commented on src/sentry/event_hashing.py:
```diff
@@ -11,13 +11,292 @@
+    """Given an event this returns a dictionary of the matching grouping
+    variants.  Checksum and fingerprinting logic are not handled by this
+    function.  Note that this will completely skip over strategies that
+    were not attempted which is suboptimal from the user experience.
```
Comment is outdated. updated.

mitsuhiko commented on src/sentry/event_hashing.py:
```diff
@@ -11,13 +11,292 @@
+    function.  Note that this will completely skip over strategies that
+    were not attempted which is suboptimal from the user experience.
+    """
+    interfaces = event.get_interfaces()
```
Done.

mitsuhiko commented on src/sentry/event_hashing.py:
```diff
@@ -11,13 +11,292 @@
+    """
+    interfaces = event.get_interfaces()
+
+    winning_strategy = None
```
Yes. The reason for this is that i'm moving all this code afterwards into a `legacy` strategy and this code will be able to try multiple strategies. So this will later likely report as something like `exception:legacy` or `exception:default`.

mitsuhiko commented on src/sentry/event_hashing.py:
```diff
@@ -11,13 +11,292 @@
+            if winning_strategy:
+                component.update(hint='%s strategy takes precedence' % winning_strategy)
+            else:
+                component.update(hint='nothing matched')
```
I think this branch is dead. Will remove.

mitsuhiko commented on src/sentry/interfaces/base.py:
```diff
@@ -198,24 +199,89 @@ def get_api_meta(self, meta, is_public=False):
-        app_hash = self.get_hash(platform, variant='app')
-        if system_hash != app_hash and app_hash:
-            hashes['app'] = app_hash
+    def get_grouping_component(self, platform=None, variant=None):
```
I kinda want to just make platform and variant mandatory. Right now variant is not mandatory because some code passes platform by position. The current explanation i made myself is that if no variant is passed then whatever is the default logic should be used. This is not too unreasonable for nested calls of interface code.

mitsuhiko commented on src/sentry/interfaces/exception.py:
```diff
@@ -1036,30 +1066,39 @@ def to_json(self):
+        # component directly
+        exceptions = self._values()
+        if len(exceptions) == 1:
+            return exceptions[0].get_grouping_component(platform, variant)
```
So this requires a bit of explanation. We have a weird setup where the `exception` key in the event payload is actually multiple exceptions made of the `SingleException` interface. However you can also supply a single exception in which case we normalize that into `[{that execption}]` now.

I retrofitted the existing grouping logic to be just a special case of the single exception grouping code. All the special handling however only applies if multiple exceptions are sent, so I can make a quick skip if there is exactly one.

mitsuhiko commented:
> I also thought that we had talked about making these changes outside of the interfaces so that they could be tested alongside the existing implementation without having to commit to changing everything at once. That doesn't seem to be the case here?

My next goal is to move this all into a new module `sentry.grouping.strategies.legacy` and then have alternative implementations side by side. I did want to move it immediately but I was worried of regressing because that would also change the tests.

Additionally I still use the separate function calls in `event_hashing` to generate the hashes instead of the components right away. If I feel confident I will remove them in the same PR.

mitsuhiko commented:
I discovered that we introduced a regression on master which was caused by unintentionally fixing a bug. The react-native test I check in shows the behavior. 688c67eb120791875a28a102adc1ae1f0311b5e6 has it working, current master (e3c6f22805b9aaaaa7ba41003946dd0e5f894340) has it broken and this branch "fixes" it again.

Effectively we forgot to pass the `system_frames` property down to the inner exceptions when building a hash in the fallback path:
 https://github.com/getsentry/sentry/blob/c59695046a41c4e93dfb6da3f9677428969a9d51/src/sentry/interfaces/exception.py#L1063-L1066

This caused the stacktrace to always be the same for system/in-app. This matters because we discard the in-app stacktrace if less than 10% of frames are in-app. By intentionally removing this behavior by passing variant down in the new code we produce fallback group now which is super generic (type name + value).

The solution I implemented to this which is not a complete hack is some postprocessing on the variants: https://github.com/getsentry/sentry/blob/90eeccc90e22568f2769bac7c977b0de3581d2c4/src/sentry/interfaces/exception.py#L799-L836

mitsuhiko commented:
I feel reasonably confident about merging this now. In any case it fixes an issue we already introduced on master and I could not find any further regressions.

mitsuhiko commented:
Test failure seems to be unrelated.

tkaemming commented on src/sentry/interfaces/exception.py:
```diff
@@ -795,6 +796,46 @@ def iter_tags(self):
+            # this branch is basically dead because we only have two
+            # variants right now, but this is so this does not break in
+            # the future.
+            hint_suffix = 'othes do'
```
typo here,
```suggestion
            hint_suffix = 'others do'
```

tkaemming commented on src/sentry/interfaces/exception.py:
```diff
@@ -795,6 +796,46 @@ def iter_tags(self):
             yield ('handled', self.handled and 'yes' or 'no')
 
 
+def uncontribute_non_stacktrace_variants(variants):
```
More specific context around this being that it's to support a legacy unintended behavior would be helpful so that we don't think that this is actually desirable for some reason, I think?

tkaemming commented on src/sentry/interfaces/exception.py:
```diff
@@ -795,6 +796,46 @@ def iter_tags(self):
+        return variants
+    any_stacktrace_contributes = False
+    non_contributing_variants = []
+    stacktrace_variants = set()
```
Kind of confusing how `non_contributing_variants` contains components but `stacktrace_variants` contains variant names, this naming could probably be improved?
